src/elf/gnu_hash.rs

```diff
@@ -47,95 +53,130 @@ macro_rules! elf_gnu_hash_impl {
         use crate::strtab::Strtab;
         use super::sym;
 
-        pub struct GnuHash<'process> {
+        const BYTE_SIZE: u32 = 8;
+        /// Size of a bits mask in bloom filter
+        const ELFCLASS_BITS: u32 = mem::size_of::<$size>() as u32 * BYTE_SIZE;
+
+        #[derive(Debug)]
+        pub struct GnuHash<'a> {
+            /// Number of GNU hash `buckets`
             nbuckets: u32,
-            symindex: usize,
+            /// Index of the first symbol in the .dynsym table which is accessible with
+            /// the hash table
+            symindex: u32,
+            /// Number of elements in the bloom filter, minus 1
+            maskwords: u32,
+            /// Shift count used in the bloom filter
             shift2: u32,
-            maskbits: u32,
-            bloomwords: &'process [$size], // either 32 or 64 bit masks, depending on platform
-            maskwords_bitmask: u32,
-            buckets: &'process [u32],
-            hashvalues: &'process [u32],
-            symtab: &'process [sym::Sym],
+            /// 2 bit bloom filter on `chains`
+            // Either 32 or 64-bit depending on the class of object
+            bloom_filter: &'a [$size], // => bloom_filter[maskwords]
+            /// GNU hash table bucket array; indexes start at 0. This array holds symbol
+            /// table indexes and contains the index of hashes in `chains`
+            buckets: &'a [u32], // => buckets[nbuckets]
+            /// Hash values; indexes start at 0. This array holds symbol table indexes.
+            chains: &'a [u32], // => chains[total_dynsyms - symindex]
         }
 
-        impl<'process> GnuHash<'process> {
-            pub unsafe fn new(hashtab: *const u32, total_dynsyms: usize, symtab: &'process [sym::Sym]) -> GnuHash<'process> {
+        impl<'a> GnuHash<'a> {
+            pub unsafe fn new(hashtab: *const u32, total_dynsyms: usize) -> GnuHash<'a> {
+                const WORD_SIZE: u32 = 32;
+                const RATIO: u32 = ELFCLASS_BITS / WORD_SIZE;
                 let nbuckets = *hashtab;
-                let symindex = *hashtab.add(1) as usize;
-                let maskwords = *hashtab.add(2) as usize; // how many words our bloom filter mask has
+                let symindex = *hashtab.add(1);
+                let maskwords = *hashtab.add(2); // how many words our bloom filter mask has
+                assert!((maskwords & (maskwords - 1)) == 0);
                 let shift2 = *hashtab.add(3);
-                let bloomwords_ptr = hashtab.add(4) as *const $size;
-                let buckets_ptr = bloomwords_ptr.add(maskwords) as *const u32;
+                #[deny(clippy::cast_ptr_alignment)]
+                let bloom_filter_ptr = hashtab.add(4) as *const $size;
+                let buckets_ptr = bloom_filter_ptr.add((RATIO * maskwords) as usize) as *const u32;
```

Comment:
Same as https://github.com/bminor/glibc/blob/94e358f6d490650c714edb1ffc3a52f56ffe086e/elf/dl-lookup.c#L953 .

Reply:
How did you test this?